### PR TITLE
chore: improve error message for bytestring length check

### PR DESCRIPTION
**Pull request by tserg**
### What I did

The following invalid syntax runs into an error when raising an exception - (and similarly for `Bytes[N]`):
```
@external
def foo():
    a: String = "abc"
```
```
TypeError: _BytestringT.compare_type() missing 1 required positional argument: 'other'
```

This should have been caught here:
https://github.com/vyperlang/vyper/blob/a5995a91b769544b43819ef68ab6be5b6c6e1274/vyper/semantics/types/bytestrings.py#L125-L129

Additionally, the error message for `DynArray` and `HashMap` can be improved:
```
@external
def foo():
    a: DynArray = [1, 2, 3]
```
```
vyper.exceptions.InvalidType: Expected <class 'vyper.semantics.types.subscriptable.DArrayT'> but literal can only be cast as DynArray[int104, 3] or uint96[3].
```

This compiles when it should raise:
```
b: HashMap
```

### How I did it

Call `from_annotation` for its side effects in `type_from_annotation` if the attribute exists for the retrieved type.

### How to verify it

See tests.

### Commit message

```
chore: improve error message for bytestring length check
```

### Description for the changelog

Improve error message for bytestring length check

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: tests/parser/syntax/test_byte_string.py [hunk 1]
```diff
@@ -1,26 +0,0 @@
...
-@external
-def foo():
-    x: String[11] = "¡très bien!"
-    """,
-    """
-@external
-def test() -> String[100]:
-    return "hello world!"
-    """,
-]
```
Comment (tserg) [inline tests/parser/syntax/test_byte_string.py:21]: Duplicated tests: https://github.com/vyperlang/vyper/blob/a5995a91b769544b43819ef68ab6be5b6c6e1274/tests/parser/syntax/test_string.py#L5-L15

I moved the last example to this file.

File: vyper/semantics/types/utils.py [hunk 2]
```diff
@@ -104,7 +104,12 @@ def _failwith(type_name):
     if node.id not in namespace:
         _failwith(node.node_source_code)
 
-    return namespace[node.id]
+    typ_ = namespace[node.id]
+    if hasattr(typ_, "from_annotation"):
```
Comment (charles-cooper) [inline vyper/semantics/types/utils.py:108]: do any types not have this method?

Comment (tserg) [reply vyper/semantics/types/utils.py:108]: Yes, only bytestrings an subscriptable types have this method.

Comment (charles-cooper) [reply vyper/semantics/types/utils.py:108]: ah, i see

File: vyper/semantics/types/utils.py [hunk 3]
```diff
@@ -104,7 +104,12 @@ def _failwith(type_name):
...
+        # for its side effects (will throw if there is a structure exception)
```
Comment (charles-cooper) [inline vyper/semantics/types/utils.py:109]:
```suggestion
        # cases where the object in the namespace is an uninstantiated
        # type object, ex. Bytestring or DynArray (with no length provided).
        # call from_annotation to produce a better error message.
```

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**